File: coding/Algorithm/BinaryTreeTraversal.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    explicit TreeNode(int x) :
            val(x), left(nullptr), right(nullptr) {
    }
};

/*
 * 遍历二叉树
 */
class BinaryTreeTraversal {
private:
    // 遍历结果
    vector<int> result;

    // 前序遍历的递归实现
    void FrontNext(TreeNode *Node){
        if (Node == nullptr) {
            return;
        }
        result.push_back(Node->val);
        FrontNext(Node->left);
        FrontNext(Node->right);
    }

    // 中序遍历的递归实现
    void MidNext(TreeNode *Node){
        if (Node == nullptr) {
            return;
        }
        MidNext(Node->left);
        result.push_back(Node->val);
        MidNext(Node->right);
    }

    // 后序遍历的递归实现
    void AfterNext(TreeNode *Node){
        if (Node == nullptr) {
            return;
        }
        AfterNext(Node->left);
        AfterNext(Node->right);
        result.push_back(Node->val);
    }

    // 前序遍历非递归实现
    void FrontOrder(TreeNode *Node){
        stack<TreeNode*> tmp;
        while (!tmp.empty() || Node) {
            if (Node) {
                result.push_back(Node->val);
                tmp.push(Node);
                Node = Node->left;
            } else {
                Node=tmp.top();
                tmp.pop();
                Node = Node->right;
            }
        }
    }

    // 中序遍历非递归实现
    void MidOrder(TreeNode *Node){
        stack<TreeNode*> tmp;
        while (!tmp.empty() || Node) {
            if (Node) {
                tmp.push(Node);
                Node = Node->left;
            } else{
                Node=tmp.top();
                tmp.pop();
                result.push_back(Node->val);
                Node = Node->right;
            }
        }
    }

    // 后序遍历非递归实现
    void AfterOrder(TreeNode *Node){
        stack<TreeNode*> tmp;
        TreeNode *last = nullptr;
        while (Node) {
            tmp.push(Node);
            Node = Node->left;
        }
        while (!tmp.empty()) {
            Node = tmp.top();
            if(Node->right== nullptr||Node->right==last){
                tmp.pop();
                result.push_back(Node->val);
                last = Node;
            } else{
                Node = Node->right;
                while (Node) {
                    tmp.push(Node);
                    Node = Node->left;
                }
            }
        }
    }

public:
    void Traversal(TreeNode *pRoot, char method = 'f', bool recursion= true){
        result.clear();
        switch (method) {
            case 'f':
                if(recursion){
                    FrontNext(pRoot);
                } else{
                    FrontOrder(pRoot);
                }
                break;
            case 'm':
                if(recursion){
                    MidNext(pRoot);
                } else {
                    MidOrder(pRoot);
                }
                break;
            case 'a':
                if(recursion){
                    AfterNext(pRoot);
                } else{
                    AfterOrder(pRoot);
                }
                break;
            default:
                return;
        }
    }

    void ShowResult(){
        for (auto elem: result) {
            cout << elem << endl;
        }
    }

    vector<int> GetOrder(){
        return result;
    }
};


int main(){
    auto *root=new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->left->left = new TreeNode(5);
    BinaryTreeTraversal so;
    cout << "RecursionTraversalFront:" << endl;
    so.Traversal(root, 'f');
    so.ShowResult();
    cout << "RecursionTraversalMid:" << endl;
    so.Traversal(root, 'm');
    so.ShowResult();
    cout << "RecursionTraversalAfter:" << endl;
    so.Traversal(root, 'a');
    so.ShowResult();
    cout << "NonRecursionTraversalFront:" << endl;
    so.Traversal(root, 'f', false);
    so.ShowResult();
    cout << "NonRecursionTraversalMid:" << endl;
    so.Traversal(root, 'm', false);
    so.ShowResult();
    cout << "NonRecursionTraversalAfter:" << endl;
    so.Traversal(root, 'a', false);
    so.ShowResult();
    return 0;
}

File: coding/java/src/BinaryTreeTraversal.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;


public class BinaryTreeTraversal {
    private LinkedList<Integer> result;

    private BinaryTreeTraversal() {
        result = new LinkedList<>();
    }

    public static void main(String[] argv) {
        TreeNode header = new TreeNode(1);
        header.left = new TreeNode(2);
        header.right = new TreeNode(3);
        header.left.left = new TreeNode(4);
        header.left.right = new TreeNode(5);
        header.right.left = new TreeNode(6);
        header.right.right = new TreeNode(7);
        BinaryTreeTraversal traversal = new BinaryTreeTraversal();
        traversal.order_after(header);
        for (int elem : traversal.result) {
            System.out.println(elem);
        }
    }

    // 递归先序遍历
    private void traversal_front(TreeNode head) {
        if (head == null) {
            return;
        }
        result.add(head.val);
        traversal_front(head.left);
        traversal_front(head.right);
    }

    // 递归中序遍历
    private void traversal_mid(TreeNode head) {
        if (head == null) {
            return;
        }
        traversal_mid(head.left);
        result.add(head.val);
        traversal_mid(head.right);
    }

    // 递归后序遍历
    private void traversal_after(TreeNode head) {
        if (head == null) {
            return;
        }
        traversal_after(head.left);
        traversal_after(head.right);
        result.add(head.val);
    }

    // 非递归前序
    private void order_front(TreeNode head) {
        Stack<TreeNode> sta = new Stack<TreeNode>();
        while (head != null) {
            result.add(head.val);
            sta.push(head);
            head = head.left;
        }
        TreeNode node;
        while (!sta.empty()) {
            node = sta.pop();
            if (node.right != null) {
                node = node.right;
                while (node != null) {
                    result.add(node.val);
                    sta.push(node);
                    node = node.left;
                }
            }
        }
    }

    // 非递归中序
    private void order_mid(TreeNode head) {
        Stack<TreeNode> sta = new Stack<TreeNode>();
        while (head != null) {
            sta.push(head);
            head = head.left;
        }
        TreeNode node;
        while (!sta.empty()) {
            node = sta.pop();
            result.add(node.val);
            if (node.right != null) {
                node = node.right;
                while (node != null) {
                    sta.push(node);
                    node = node.left;
                }
            }
        }
    }

    // 非递归后续遍历
    private void order_after(TreeNode head) {
        Stack<TreeNode> sta = new Stack<TreeNode>();
        while (head != null) {
            sta.push(head);
            head = head.left;
        }
        TreeNode node;
        TreeNode last = new TreeNode(-9999);
        while (!sta.empty()) {
            node = sta.peek();
            if (node.right == null || node.right == last) {
                result.add(node.val);
                sta.pop();
                last = node;
            } else {
                node = node.right;
                while (node != null) {
                    sta.push(node);
                    node = node.left;
                }
            }
        }
    }
}

